### DISCUSSION TITLE
2.0 migration - can't `subqueryload` after a `joinedload`, vice versa

### DISCUSSION THREAD

**Discussion by jvanasco** [sqlalchemy/sqlalchemy#10430] [Usage Questions] [answered]
I'm in the middle of migrating a large 1.4 project to 2.0

Under the old "string" syntax for eager loading, I could essentially do the following to control eager loading selects:

    query = query.options(joinedload('bar').joinedload('bar.biz')
    query = query.options(subqueryload('bar').joinedload('bar.bash')

This is no longer possible with the removal of strings as a loading option.

The following will generate an exception `sqlalchemy.exc.InvalidRequestError: Loader strategies for ORM Path[Mapper[Foo(foo)] -> Foo.bar -> Mapper[Bar(bar)]] conflict` as the initial paths do not match:

    query = query.options(joinedload(Foo.bar).joinedload(Bar.biz)
    query = query.options(subqueryload(Foo.bar).joinedload(Bar.bash)

Is there anything I can do to try and approximate the old behavior, without massively restructuring code?

The project I am migrating has a lot of complex eager loads that have been tuned for subquery vs joined loading.  

```python
"""
sqlatest.py

works:
    python sqlatest.py

fails:
    python sqlatest.py x
"""
# stdlib
import sys

# pypi
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy.orm import configure_mappers
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import joinedload
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session
from sqlalchemy.orm import subqueryload

# Note: env
TRIGGER = True if len(sys.argv) > 1 else False

# Note: Model


class Base(DeclarativeBase):
    pass


class Foo(Base):
    __tablename__ = "foo"
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    bar_id: Mapped[int] = mapped_column(Integer, ForeignKey("bar.id"), nullable=True)
    bar = relationship("Bar", primaryjoin="Foo.bar_id==Bar.id")


class Bar(Base):
    __tablename__ = "bar"
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    bar_biz_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("bar_biz.id"), nullable=True
    )
    bar_bash_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("bar_bash.id"), nullable=True
    )
    bar_biz = relationship("BarBiz", primaryjoin="Bar.bar_biz_id==BarBiz.id")
    bar_bash = relationship("BarBash", primaryjoin="Bar.bar_bash_id==BarBash.id")


class BarBiz(Base):
    __tablename__ = "bar_biz"
    id: Mapped[int] = mapped_column(Integer, primary_key=True)


class BarBash(Base):
    __tablename__ = "bar_bash"
    id: Mapped[int] = mapped_column(Integer, primary_key=True)


# Note: Initialize Database

configure_mappers()

engine = create_engine("sqlite://", echo=True)
Base.metadata.create_all(engine)

# Note: Query
dbSession = Session(engine)
query = dbSession.query(Foo).filter(Foo.id == 1)
query = query.options(subqueryload(Foo.bar).joinedload(Bar.bar_biz))
if TRIGGER:
    query = query.options(joinedload(Foo.bar).joinedload(Bar.bar_bash))
else:
    query = query.options(subqueryload(Foo.bar).joinedload(Bar.bar_bash))
item = query.first()
```

**Comment by zzzeek** [comment]
So I didnt understand your original description, and I dont understand what you want the code to do:

```
query = query.options(subqueryload(Foo.bar).joinedload(Bar.bar_biz))
if TRIGGER:
    query = query.options(joinedload(Foo.bar).joinedload(Bar.bar_bash))
else:
    query = query.options(subqueryload(Foo.bar).joinedload(Bar.bar_bash))
```

if TRIGGER, do you expect the Foo.bar to be a joinedload and the subqueryload goes away?   the error message shown here is a new feature added in 2.0 to resolve for this ambiguity.

for 2.0, you'd want to restate this code so that the Foo.bar loader path is given an option only once, using `defaultload()` for the path that wants to inherit whatever the final decision is:

```
query = query.options(defaultload(Foo.bar).joinedload(Bar.bar_biz))

if TRIGGER:
    query = query.options(joinedload(Foo.bar).joinedload(Bar.bar_bash))
else:
    query = query.options(subqueryload(Foo.bar).joinedload(Bar.bar_bash))


```

**Reply by jvanasco** [reply]
Thanks for the quick reply, Mike.

Sorry for the confusion - this is a very stripped down example – and this **might** only be triggered by a new test harness written to test the 2.0 migration.  

It was my assumption SQLAlchemy's planner was not trying to merge the different strategies and the code would emit multiple select statements - one for the subqueryload and the other for the joinedload.

**Reply by zzzeek** [reply]
it only loads a particular relationship, in this case `Foo.bar` , using **one** strategy.

otherwise, what do you think this would do?

```py
stmt = select(Foo).options(selectinload(Foo.bars), subqueryload(Foo.bars))
```

**Reply by zzzeek** [reply, marked as answer]
if the command you are sending out only cares about `Bar.bar_bash` , then use `defaultload()` in that way:

```
query = query.options(joinedload(Foo.bar).joinedload(Bar.bar_biz))

if TRIGGER:
    query = query.options(defaultload(Foo.bar).joinedload(Bar.bar_bash))
else:
    query = query.options(defaultload(Foo.bar).subqueryload(Bar.bar_bash))

```

above, you will get `Bar.bar_bash` either via join or by subqueryload.   Foo.bar and Bar.bar_biz always by joined.

**Reply by jvanasco** [reply]
Thanks, @zzzeek - I think you **exactly** determined what I am trying to accomplish here.

**Reply by CaselIT** [reply]
so in that case `defaultload` does not imply the default loading strategy but the currently configured one, falling back on the default one if none was set, correct?

if so maybe the docs could be clarified a bit?

**Comment by jvanasco** [comment]
> so in that case defaultload does not imply the default loading strategy but the currently configured one, falling back on the default one if none was set, correct?

Unless I have messed up my test case, it seems to do the following, in order:

1- Use the most recent strategy.  

    query = query.options(joinedload(Foo.bar).joinedload(Bar.bar_biz))
    query = query.options(defaultload(Foo.bar).joinedload(Bar.bar_bash))

The  `defaultload` will be `joinedload`

2- Use the default specified in the relationship configuraiton
3- Not do anything, meaning no query will take place.

The third bit should probably raise an exception

**Reply by CaselIT** [reply]
> 3- Not do anything, meaning no query will take place.

> The third bit should probably raise an exception

I don't think it should. If you have this case
```py
# Foo.bar uses the default lazy='select'
query = query.options(defaultload(Foo.bar).joinedload(Bar.bar_bash))
```
you can use it to indicate that `Foo.bar` will lazy load, but when you load it `Bar.bar_bash` will load it with joined load

**Reply by zzzeek** [reply]
defaultload() means "change nothing" for that particular element of the load path.  it's used to give a launching point for the next link in the path and nothing else.

**Reply by CaselIT** [reply]
ok, I'll write something on the docs

**Reply by CaselIT** [reply]
let me know if it's any better https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4880

**Comment by jvanasco** [comment]
Ah, that makes sense. Nevermind.  The test harness I had didn't try to access the attribute, it just monitored the engine's sql output.
